File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/metrics/Timer.java
package com.github.dynamicextensionsalfresco.metrics;

import java.util.function.Supplier;

/**
 * register timing information during a transaction and report after commit (reporting dependant of implementation)
 *
 * @author Laurent Van der Linden
 */
public interface Timer {

    void start(String label);

    void stop();

    boolean isEnabled();

    default <T, E extends Throwable> T time(Supplier<String> labelProvider, ThrowingSupplier<T, E> operation) throws E {
        if (isEnabled()) {
            start(labelProvider.get());
            try {
                return operation.get();
            } finally {
                stop();
            }
        } else {
            return operation.get();
        }
    }

    @FunctionalInterface
    interface ThrowingSupplier<T, E extends Throwable> {

        T get() throws E;
    }
}

File: annotations-runtime/src/main/kotlin/com/github/dynamicextensionsalfresco/metrics/Timer.kt
package com.github.dynamicextensionsalfresco.metrics

/**
 * register timing information during a transaction and report after commit
 * (reporting dependant of implementation)
 *
 * @author Laurent Van der Linden
 */
public interface Timer {
    fun start(label: String)

    fun stop()

    val enabled: Boolean
}

inline fun Timer.time(labelProvider: () -> String, operation: () -> Unit) {
    if (enabled) {
        start(labelProvider.invoke())
        try {
            operation()
        } finally {
            stop()
        }
    } else {
        operation()
    }
}
